stdlib/public/core/String.swift

```diff
@@ -470,16 +502,34 @@ extension String : Hashable {
 #else
     let hashOffset = Int(bitPattern: 0x429b_1266_88dd_cc21)
 #endif
-    // FIXME(performance): constructing a temporary NSString is extremely
-    // wasteful and inefficient.
-    let cocoaString = unsafeBitCast(
-      self._bridgeToObjectiveCImpl(), to: _NSStringCore.self)
-
-    // If we have an ASCII string, we do not need to normalize.
-    if self._core.isASCII {
-      return hashOffset ^ _stdlib_NSStringASCIIHashValue(cocoaString)
+    // If we have a contigous string then we can use the stack optimization.
+    let core = self._core
+    if core.hasContiguousStorage {
+      // If we have an ASCII string, we do not need to normalize.
+      if core.isASCII {
+        let res = hashOffset ^ _stdlib_NSStringASCIIHashValueStack(
+          core._baseAddress,
+          core._countAndFlags,
+          unsafeBitCast(core._owner, to: OpaquePointer.self))
+        _fixLifetime(self)
+        return res
+      } else {
+        let res = hashOffset ^ _stdlib_NSStringNFDHashValueStack(
+          core._baseAddress,
+          core._countAndFlags,
+          unsafeBitCast(core._owner, to: OpaquePointer.self))
+        _fixLifetime(self)
+        return res
```

Comment:
Instead of doubling the number of functions (and increasing the amount of code at the call site), can we make `isASCII` a function parameter?